File: example_usage.py
"""
Practical Examples: Using the Unified Decision Interface
Demonstrates real-world scenarios across different hardware
"""

from decision_interface import *
import numpy as np
import time


def example_1_smart_firewall():
    """
    Smart Firewall: Use OCTEON for packet filtering + Stingray for ML threat detection
    """
    print("\n" + "="*70)
    print("EXAMPLE 1: Smart Firewall with Hybrid Processing")
    print("="*70)

    interface = UnifiedDecisionInterface()
    interface.register_engine(OcteonEngine({"num_cores": 24, "packet_engines": 8}))
    interface.register_engine(StingrayEngine({"num_cores": 32}))

    # Scenario: Incoming network traffic
    packets = [
        InputData(
            raw_data=b"GET /api/users HTTP/1.1",
            metadata={"src_ip": "192.168.1.100", "dst_port": 443},
            data_format="packet"
        ),
        InputData(
            raw_data=b"malicious SQL injection attempt",
            metadata={"src_ip": "10.0.0.50", "dst_port": 3306},
            data_format="packet"
        ),
        InputData(
            raw_data=b"authorized admin session",
            metadata={"src_ip": "192.168.1.1", "dst_port": 22},
            data_format="packet"
        ),
    ]

    print("\nProcessing packets through OCTEON DPU...")
    for i, packet in enumerate(packets, 1):
        # Fast hardware-based classification
        decision = interface.process(packet, HardwareType.OCTEON)

        print(f"\nPacket {i}:")
        print(f"  Source: {packet.metadata['src_ip']}")
        print(f"  Action: {decision.result}")
        print(f"  Confidence: {decision.confidence:.2%}")
        print(f"  Latency: {decision.latency_ms:.3f}ms")

        # If suspicious, escalate to ML analysis on ARM
        if decision.result == "inspect":
            print(f"  → Escalating to Stingray for deep ML analysis...")
            ml_decision = interface.process(packet, HardwareType.STINGRAY)
            print(f"  → ML Result: {ml_decision.result}")

    interface.shutdown_all()


def example_2_autonomous_vehicle():
    """
    Autonomous Vehicle: Multi-sensor fusion across heterogeneous hardware
    """
    print("\n" + "="*70)
    print("EXAMPLE 2: Autonomous Vehicle Multi-Sensor Processing")
    print("="*70)

    interface = UnifiedDecisionInterface()
    interface.register_engine(StingrayEngine({"num_cores": 32}))
    interface.register_engine(TrueNorthEngine({"num_cores": 4096}))

    # Simulate sensors
    print("\nProcessing multi-modal sensor data...")

    # Camera feed → Stingray (CNN for object detection)
    camera_frame = InputData(
        raw_data=np.random.rand(640, 480, 3),
        metadata={"sensor": "front_camera", "timestamp": time.time()},
        data_format="tensor"
    )
    vision_decision = interface.process(camera_frame, HardwareType.STINGRAY)
    print(f"\n[Camera] Object detected: Class {vision_decision.result}")
    print(f"         Confidence: {vision_decision.confidence:.2%}")
    print(f"         Latency: {vision_decision.latency_ms:.2f}ms")

    # Event-based sensor → TrueNorth (neuromorphic processing)
    event_stream = InputData(
        raw_data=np.random.rand(1024),
        metadata={"sensor": "event_camera", "timestamp": time.time()},
        data_format="spike_train"
    )
    event_decision = interface.process(event_stream, HardwareType.TRUENORTH)
    print(f"\n[Event Sensor] Pattern: {event_decision.result}")
    print(f"               Confidence: {event_decision.confidence:.2%}")
    print(f"               Power: {event_decision.metadata['power_mw']}mW")
    print(f"               Latency: {event_decision.latency_ms:.2f}ms")

    # Decision fusion
    print("\n[Fusion] Combining decisions from multiple modalities...")
    print(f"         → Vision confidence: {vision_decision.confidence:.2%}")
    print(f"         → Event confidence: {event_decision.confidence:.2%}")
    print(f"         → Final decision: SAFE TO PROCEED")

    interface.shutdown_all()


def example_3_edge_ai_gateway():
    """
    Edge AI Gateway: Route different workloads to optimal hardware
    """
    print("\n" + "="*70)
    print("EXAMPLE 3: Edge AI Gateway with Intelligent Routing")
    print("="*70)

    interface = UnifiedDecisionInterface()
    interface.register_engine(StingrayEngine({"num_cores": 32}))
    interface.register_engine(OcteonEngine({"num_cores": 24, "packet_engines": 8}))
    interface.register_engine(TrueNorthEngine({"num_cores": 4096}))

    # Various input types
    inputs = [
        InputData(
            raw_data=np.random.rand(224, 224, 3),
            metadata={"type": "image", "source": "security_camera"},
            data_format="tensor"
        ),
        InputData(
            raw_data=b"network packet data",
            metadata={"type": "packet", "protocol": "tcp"},
            data_format="packet"
        ),
        InputData(
            raw_data=np.random.rand(512),
            metadata={"type": "sensor", "source": "vibration_monitor"},
            data_format="spike_train"
        ),
    ]

    print("\nAutomatically routing workloads to optimal hardware...")

    for i, input_data in enumerate(inputs, 1):
        # Automatically select best hardware
        best_hw = interface.get_best_hardware(input_data)
        decision = interface.process(input_data, best_hw)

        print(f"\nInput {i}: {input_data.metadata['type']}")
        print(f"  Routed to: {decision.hardware_type.value}")
        print(f"  Result: {decision.result}")
        print(f"  Latency: {decision.latency_ms:.3f}ms")
        print(f"  Metadata: {decision.metadata}")

    interface.shutdown_all()


def example_4_batch_processing():
    """
    Batch Processing: Maximize throughput with batched operations
    """
    print("\n" + "="*70)
    print("EXAMPLE 4: High-Throughput Batch Processing")
    print("="*70)

    interface = UnifiedDecisionInterface()
    interface.register_engine(StingrayEngine({"num_cores": 32}))
    interface.register_engine(OcteonEngine({"num_cores": 24, "packet_engines": 8}))

    # Generate large batch of data
    batch_sizes = [10, 100, 1000]

    for batch_size in batch_sizes:
        print(f"\n--- Batch Size: {batch_size} ---")

        # Image batch on Stingray
        image_batch = [
            InputData(
                raw_data=np.random.rand(224, 224, 3),
                metadata={"batch_idx": i},
                data_format="tensor"
            )
            for i in range(batch_size)
        ]

        start = time.time()
        decisions = interface.batch_process(image_batch, HardwareType.STINGRAY)
        stingray_time = time.time() - start

        print(f"Stingray (ARM): {batch_size} images in {stingray_time*1000:.2f}ms")
        print(f"  Throughput: {batch_size/stingray_time:.0f} inferences/sec")

        # Packet batch on OCTEON
        packet_batch = [
            InputData(
                raw_data=b"packet data",
                metadata={"batch_idx": i},
                data_format="packet"
            )
            for i in range(batch_size)
        ]

        start = time.time()
        decisions = interface.batch_process(packet_batch, HardwareType.OCTEON)
        octeon_time = time.time() - start

        print(f"OCTEON (DPU): {batch_size} packets in {octeon_time*1000:.2f}ms")
        print(f"  Throughput: {batch_size/octeon_time:.0f} packets/sec")

    interface.shutdown_all()


def example_5_hardware_capabilities():
    """
    Hardware Introspection: Query capabilities and constraints
    """
    print("\n" + "="*70)
    print("EXAMPLE 5: Hardware Capability Discovery")
    print("="*70)

    interface = UnifiedDecisionInterface()
    interface.register_engine(StingrayEngine({"num_cores": 32}))
    interface.register_engine(OcteonEngine({"num_cores": 24, "packet_engines": 8}))
    interface.register_engine(TrueNorthEngine({"num_cores": 4096}))

    print("\nQuerying hardware capabilities...\n")

    for hw_type, engine in interface.engines.items():
        caps = engine.get_capabilities()

        print(f"{hw_type.value.upper()}")
        print("-" * 50)
        for key, value in caps.items():
            print(f"  {key:25s}: {value}")
        print()

    interface.shutdown_all()


def example_6_fallback_strategy():
    """
    Fallback Strategy: Handle hardware failures gracefully
    """
    print("\n" + "="*70)
    print("EXAMPLE 6: Hardware Fallback and Resilience")
    print("="*70)

    interface = UnifiedDecisionInterface()
    interface.register_engine(StingrayEngine({"num_cores": 32}))
    interface.register_engine(OcteonEngine({"num_cores": 24}))

    # Try processing with fallback
    input_data = InputData(
        raw_data=np.random.rand(224, 224, 3),
        metadata={"type": "image"},
        data_format="tensor"
    )

    # Preferred order: Try TrueNorth first (not registered), fall back to Stingray
    print("\nAttempting processing with fallback strategy...")
    print("Preferred order: TrueNorth → OCTEON → Stingray")

    try:
        decision = interface.process_with_fallback(
            input_data,
            [HardwareType.TRUENORTH, HardwareType.OCTEON, HardwareType.STINGRAY]
        )
        print(f"\nSuccessfully processed on: {decision.hardware_type.value}")
        print(f"Result: {decision.result}")
        print(f"Latency: {decision.latency_ms:.2f}ms")
    except Exception as e:
        print(f"\nError: {e}")

    interface.shutdown_all()


def example_7_power_efficiency():
    """
    Power Efficiency: Compare energy consumption across hardware
    """
    print("\n" + "="*70)
    print("EXAMPLE 7: Power Efficiency Comparison")
    print("="*70)

    interface = UnifiedDecisionInterface()
    interface.register_engine(StingrayEngine({"num_cores": 32}))
    interface.register_engine(TrueNorthEngine({"num_cores": 4096}))

    input_data = InputData(
        raw_data=np.random.rand(1024),
        metadata={"type": "pattern"},
        data_format="tensor"
    )

    print("\nComparing power consumption for same workload...\n")

    # Process on Stingray
    decision_arm = interface.process(input_data, HardwareType.STINGRAY)
    estimated_power_arm = 80  # watts (typical ARM server)

    print(f"Stingray (ARM):")
    print(f"  Latency: {decision_arm.latency_ms:.2f}ms")
    print(f"  Power: ~{estimated_power_arm}W")
    print(f"  Energy: {estimated_power_arm * decision_arm.latency_ms / 1000:.2f} mJ")

    # Process on TrueNorth
    decision_neuro = interface.process(input_data, HardwareType.TRUENORTH)
    power_neuro = float(decision_neuro.metadata.get('power_mw', 70))

    print(f"\nTrueNorth (Neuromorphic):")
    print(f"  Latency: {decision_neuro.latency_ms:.2f}ms")
    print(f"  Power: {power_neuro}mW")
    print(f"  Energy: {power_neuro * decision_neuro.latency_ms / 1000:.2f} mJ")

    energy_savings = ((estimated_power_arm * 1000) - power_neuro) / (estimated_power_arm * 1000)
    print(f"\n→ TrueNorth uses {energy_savings:.1%} less power for this workload")

    interface.shutdown_all()


if __name__ == "__main__":
    print("\n")
    print("="*70)
    print("  UNIFIED DECISION INTERFACE - PRACTICAL EXAMPLES")
    print("  Heterogeneous Hardware Abstraction Demo")
    print("="*70)

    # Run all examples
    example_1_smart_firewall()
    example_2_autonomous_vehicle()
    example_3_edge_ai_gateway()
    example_4_batch_processing()
    example_5_hardware_capabilities()
    example_6_fallback_strategy()
    example_7_power_efficiency()

    print("\n" + "="*70)
    print("All examples completed successfully!")
    print("="*70 + "\n")

File: example_usage.cpp
/**
 * Practical C++ Example: Using the Unified Decision Interface
 * Compile: g++ -std=c++17 example_usage.cpp -o decision_example
 */

#include "decision_interface.hpp"
#include <iostream>
#include <iomanip>
#include <vector>
#include <chrono>

using namespace decision;

void print_header(const std::string& title) {
    std::cout << "\n" << std::string(70, '=') << "\n";
    std::cout << title << "\n";
    std::cout << std::string(70, '=') << "\n";
}

void print_decision(const Decision& decision) {
    std::cout << "  Decision ID: " << decision.decision_id << "\n";
    std::cout << "  Hardware: " << (decision.hardware_type == HardwareType::STINGRAY ? "Stingray" :
                                    decision.hardware_type == HardwareType::OCTEON ? "OCTEON" :
                                    "TrueNorth") << "\n";
    std::cout << "  Confidence: " << std::fixed << std::setprecision(2)
              << (decision.confidence * 100) << "%\n";
    std::cout << "  Latency: " << std::fixed << std::setprecision(3)
              << decision.latency_ms << "ms\n";

    // Print metadata
    if (!decision.metadata.empty()) {
        std::cout << "  Metadata:\n";
        for (const auto& [key, value] : decision.metadata) {
            std::cout << "    " << key << ": " << value << "\n";
        }
    }
}

void example_1_smart_network() {
    print_header("EXAMPLE 1: Smart Network Processing");

    // Initialize interface
    UnifiedDecisionInterface interface;

    auto octeon = std::make_shared<OcteonEngine>(
        std::unordered_map<std::string, std::string>{
            {"num_cores", "24"},
            {"packet_engines", "8"}
        }
    );
    auto stingray = std::make_shared<StingrayEngine>(
        std::unordered_map<std::string, std::string>{{"num_cores", "32"}}
    );

    interface.register_engine(octeon, true);
    interface.register_engine(stingray);

    std::cout << "\nProcessing network packets through OCTEON DPU...\n";

    // Simulate different packet types
    std::vector<InputData> packets = {
        {
            .raw_data = std::vector<uint8_t>{'G', 'E', 'T', ' ', '/', 'a', 'p', 'i'},
            .metadata = {{"src_ip", "192.168.1.100"}, {"dst_port", "443"}},
            .data_format = "packet"
        },
        {
            .raw_data = std::vector<uint8_t>{'m', 'a', 'l', 'i', 'c', 'i', 'o', 'u', 's'},
            .metadata = {{"src_ip", "10.0.0.50"}, {"dst_port", "3306"}},
            .data_format = "packet"
        }
    };

    for (size_t i = 0; i < packets.size(); ++i) {
        std::cout << "\nPacket " << (i + 1) << ":\n";
        auto decision = interface.process(packets[i], HardwareType::OCTEON);
        print_decision(decision);

        // Escalate suspicious packets to ML analysis
        if (decision.get_result<std::string>() == "inspect") {
            std::cout << "\n  → Escalating to Stingray for deep ML analysis...\n";
            auto ml_decision = interface.process(packets[i], HardwareType::STINGRAY);
            std::cout << "  → ML Result: " << ml_decision.get_result<int>() << "\n";
        }
    }

    interface.shutdown_all();
}

void example_2_multimodal_processing() {
    print_header("EXAMPLE 2: Multi-Modal Sensor Fusion");

    UnifiedDecisionInterface interface;

    auto stingray = std::make_shared<StingrayEngine>(
        std::unordered_map<std::string, std::string>{{"num_cores", "32"}}
    );
    auto truenorth = std::make_shared<TrueNorthEngine>(
        std::unordered_map<std::string, std::string>{{"num_cores", "4096"}}
    );

    interface.register_engine(stingray, true);
    interface.register_engine(truenorth);

    std::cout << "\nProcessing multi-modal sensor data...\n";

    // Camera frame (traditional ML on ARM)
    std::cout << "\n[Camera Feed]\n";
    InputData camera_data{
        .raw_data = std::vector<float>(640 * 480 * 3, 0.5f),
        .metadata = {{"sensor", "front_camera"}, {"type", "rgb"}},
        .data_format = "tensor"
    };
    auto vision_decision = interface.process(camera_data, HardwareType::STINGRAY);
    print_decision(vision_decision);

    // Event sensor (neuromorphic processing)
    std::cout << "\n[Event Sensor]\n";
    InputData event_data{
        .raw_data = std::vector<float>(1024, 0.3f),
        .metadata = {{"sensor", "event_camera"}, {"type", "spikes"}},
        .data_format = "spike_train"
    };
    auto event_decision = interface.process(event_data, HardwareType::TRUENORTH);
    print_decision(event_decision);

    std::cout << "\n[Decision Fusion]\n";
    std::cout << "  Combined confidence: "
              << std::fixed << std::setprecision(2)
              << ((vision_decision.confidence + event_decision.confidence) / 2.0 * 100)
              << "%\n";

    interface.shutdown_all();
}

void example_3_batch_throughput() {
    print_header("EXAMPLE 3: High-Throughput Batch Processing");

    UnifiedDecisionInterface interface;

    auto stingray = std::make_shared<StingrayEngine>(
        std::unordered_map<std::string, std::string>{{"num_cores", "32"}}
    );
    auto octeon = std::make_shared<OcteonEngine>(
        std::unordered_map<std::string, std::string>{
            {"num_cores", "24"},
            {"packet_engines", "8"}
        }
    );

    interface.register_engine(stingray, true);
    interface.register_engine(octeon);

    std::vector<int> batch_sizes = {10, 100, 1000};

    for (int batch_size : batch_sizes) {
        std::cout << "\n--- Batch Size: " << batch_size << " ---\n";

        // Create batch
        std::vector<InputData> batch;
        for (int i = 0; i < batch_size; ++i) {
            batch.push_back({
                .raw_data = std::vector<float>(224 * 224 * 3, 0.5f),
                .metadata = {{"batch_idx", std::to_string(i)}},
                .data_format = "tensor"
            });
        }

        // Measure throughput
        auto start = std::chrono::high_resolution_clock::now();
        auto decisions = interface.batch_process(batch, HardwareType::STINGRAY);
        auto end = std::chrono::high_resolution_clock::now();

        double elapsed_ms = std::chrono::duration<double, std::milli>(end - start).count();
        double throughput = batch_size / (elapsed_ms / 1000.0);

        std::cout << "Stingray (ARM): " << batch_size << " images in "
                  << std::fixed << std::setprecision(2) << elapsed_ms << "ms\n";
        std::cout << "  Throughput: " << std::fixed << std::setprecision(0)
                  << throughput << " inferences/sec\n";
    }

    interface.shutdown_all();
}

void example_4_hardware_capabilities() {
    print_header("EXAMPLE 4: Hardware Capability Discovery");

    UnifiedDecisionInterface interface;

    auto stingray = std::make_shared<StingrayEngine>(
        std::unordered_map<std::string, std::string>{{"num_cores", "32"}}
    );
    auto octeon = std::make_shared<OcteonEngine>(
        std::unordered_map<std::string, std::string>{
            {"num_cores", "24"},
            {"packet_engines", "8"}
        }
    );
    auto truenorth = std::make_shared<TrueNorthEngine>(
        std::unordered_map<std::string, std::string>{{"num_cores", "4096"}}
    );

    interface.register_engine(stingray, true);
    interface.register_engine(octeon);
    interface.register_engine(truenorth);

    std::cout << "\nQuerying hardware capabilities...\n";

    for (const auto& [hw_type, engine] : interface.get_engines()) {
        std::string hw_name = (hw_type == HardwareType::STINGRAY ? "STINGRAY" :
                               hw_type == HardwareType::OCTEON ? "OCTEON 10" :
                               "TRUENORTH");

        std::cout << "\n" << hw_name << "\n";
        std::cout << std::string(50, '-') << "\n";

        auto caps = engine->get_capabilities();
        for (const auto& [key, value] : caps) {
            std::cout << "  " << std::left << std::setw(25) << key << ": " << value << "\n";
        }
    }

    interface.shutdown_all();
}

void example_5_automatic_routing() {
    print_header("EXAMPLE 5: Automatic Hardware Selection");

    UnifiedDecisionInterface interface;

    auto stingray = std::make_shared<StingrayEngine>(
        std::unordered_map<std::string, std::string>{{"num_cores", "32"}}
    );
    auto octeon = std::make_shared<OcteonEngine>(
        std::unordered_map<std::string, std::string>{
            {"num_cores", "24"},
            {"packet_engines", "8"}
        }
    );
    auto truenorth = std::make_shared<TrueNorthEngine>(
        std::unordered_map<std::string, std::string>{{"num_cores", "4096"}}
    );

    interface.register_engine(stingray, true);
    interface.register_engine(octeon);
    interface.register_engine(truenorth);

    std::cout << "\nAutomatically routing workloads to optimal hardware...\n";

    // Different input types
    std::vector<InputData> inputs = {
        {
            .raw_data = std::vector<float>(224 * 224 * 3, 0.5f),
            .metadata = {{"type", "image"}},
            .data_format = "tensor"
        },
        {
            .raw_data = std::vector<uint8_t>{'p', 'a', 'c', 'k', 'e', 't'},
            .metadata = {{"type", "network"}},
            .data_format = "packet"
        },
        {
            .raw_data = std::vector<float>(1024, 0.3f),
            .metadata = {{"type", "sensor"}},
            .data_format = "spike_train"
        }
    };

    for (size_t i = 0; i < inputs.size(); ++i) {
        std::cout << "\nInput " << (i + 1) << " (" << inputs[i].data_format << "):\n";

        // Automatic hardware selection
        auto best_hw = interface.get_best_hardware(inputs[i]);
        auto decision = interface.process(inputs[i], best_hw);

        std::string hw_name = (decision.hardware_type == HardwareType::STINGRAY ? "Stingray" :
                               decision.hardware_type == HardwareType::OCTEON ? "OCTEON 10" :
                               "TrueNorth");

        std::cout << "  Routed to: " << hw_name << "\n";
        std::cout << "  Latency: " << std::fixed << std::setprecision(3)
                  << decision.latency_ms << "ms\n";
    }

    interface.shutdown_all();
}

void example_6_power_comparison() {
    print_header("EXAMPLE 6: Power Efficiency Analysis");

    UnifiedDecisionInterface interface;

    auto stingray = std::make_shared<StingrayEngine>(
        std::unordered_map<std::string, std::string>{{"num_cores", "32"}}
    );
    auto truenorth = std::make_shared<TrueNorthEngine>(
        std::unordered_map<std::string, std::string>{{"num_cores", "4096"}}
    );

    interface.register_engine(stingray, true);
    interface.register_engine(truenorth);

    InputData input{
        .raw_data = std::vector<float>(1024, 0.5f),
        .metadata = {{"type", "pattern"}},
        .data_format = "tensor"
    };

    std::cout << "\nComparing power consumption for identical workload...\n";

    // ARM processing
    std::cout << "\nStingray (ARM):\n";
    auto arm_decision = interface.process(input, HardwareType::STINGRAY);
    double arm_power_watts = 80.0;  // Typical ARM server power
    double arm_energy_mj = arm_power_watts * 1000 * arm_decision.latency_ms / 1000.0;

    std::cout << "  Latency: " << std::fixed << std::setprecision(2)
              << arm_decision.latency_ms << "ms\n";
    std::cout << "  Power: ~" << arm_power_watts << "W\n";
    std::cout << "  Energy: " << std::fixed << std::setprecision(2)
              << arm_energy_mj << " mJ\n";

    // Neuromorphic processing
    std::cout << "\nTrueNorth (Neuromorphic):\n";
    auto neuro_decision = interface.process(input, HardwareType::TRUENORTH);
    double neuro_power_mw = std::stod(neuro_decision.metadata["power_mw"]);
    double neuro_energy_mj = neuro_power_mw * neuro_decision.latency_ms / 1000.0;

    std::cout << "  Latency: " << std::fixed << std::setprecision(2)
              << neuro_decision.latency_ms << "ms\n";
    std::cout << "  Power: " << std::fixed << std::setprecision(2)
              << neuro_power_mw << "mW\n";
    std::cout << "  Energy: " << std::fixed << std::setprecision(2)
              << neuro_energy_mj << " mJ\n";

    double power_savings = ((arm_power_watts * 1000) - neuro_power_mw) / (arm_power_watts * 1000);
    std::cout << "\n→ TrueNorth uses " << std::fixed << std::setprecision(1)
              << (power_savings * 100) << "% less power\n";

    interface.shutdown_all();
}

int main() {
    std::cout << "\n" << std::string(70, '=') << "\n";
    std::cout << "  UNIFIED DECISION INTERFACE - C++ EXAMPLES\n";
    std::cout << "  Heterogeneous Hardware Abstraction Demo\n";
    std::cout << std::string(70, '=') << "\n";

    try {
        example_1_smart_network();
        example_2_multimodal_processing();
        example_3_batch_throughput();
        example_4_hardware_capabilities();
        example_5_automatic_routing();
        example_6_power_comparison();

        std::cout << "\n" << std::string(70, '=') << "\n";
        std::cout << "All examples completed successfully!\n";
        std::cout << std::string(70, '=') << "\n\n";

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }

    return 0;
}
